File: packages/repo-gc-python/python/repo_gc_python/heuristics/unused_imports.py
"""Zombie import detection — mirrors Rust's unused_imports.rs.

Handles `from __future__ import annotations` (PEP 563) via the parsing module's
pre-processed identifier collection, which already parses annotation strings for
identifier names when the future import is active.
"""

from ..cli import Threshold
from ..discovery import PythonFile
from ..parsing import FileInfo
from ..types import Finding, FindingKind, Severity

MIN_UNUSED = 2

# Names commonly referenced by compile-time transforms rather than user code.
# Empty for Python — the language has no JSX-like implicit usage.
COMPILER_NAMES: frozenset[str] = frozenset()


def analyze(
    pfile: PythonFile,
    info: FileInfo,
    _threshold: Threshold,
    counter: int,
) -> Finding | None:
    unused = [
        name
        for name in info.import_leaf_names
        if name not in COMPILER_NAMES and name not in info.all_identifiers
    ]

    if len(unused) < MIN_UNUSED:
        return None

    severity = Severity.Medium if len(unused) >= 5 else Severity.Low

    preview = unused[:5]
    return Finding(
        id=f"ui-{counter:03d}",
        kind=FindingKind.UnusedImport,
        severity=severity,
        confidence=0.65,
        path=str(pfile.relative_path),
        summary="",
        reasons=[],
        evidence=[
            f"unused_count: {len(unused)}",
            f"preview: {', '.join(preview)}",
        ],
        suggested_next_step="",
        estimated_tokens=None,
    )

File: packages/repo-gc-shared/src/heuristics/unused-imports.ts
import { Finding, FindingKind, Severity, type FileData } from '../types';

export interface UnusedImportsOptions {
  /** Names commonly referenced by compile-time transforms rather than user code. */
  compilerNames?: Set<string>;
}

export function analyzeUnusedImports(
  data: FileData,
  idCounter: { value: number },
  options: UnusedImportsOptions = {},
): Finding | null {
  const compilerNames = options.compilerNames ?? new Set<string>();

  const unusedNames: string[] = [];
  for (const name of data.importedNames) {
    if (compilerNames.has(name)) continue;
    if (data.allIdentifiers.has(name)) continue;
    unusedNames.push(name);
  }

  // Require at least 2 unused names to reduce false positives from macros/transforms
  if (unusedNames.length < 2) return null;

  const severity = unusedNames.length >= 5 ? Severity.Medium : Severity.Low;

  return {
    id: `ui-${String(idCounter.value++).padStart(3, '0')}`,
    kind: FindingKind.UnusedImport,
    severity,
    confidence: 0.65,
    path: data.relativePath,
    summary: '',
    reasons: [],
    evidence: [
      `unused_count: ${unusedNames.length}`,
      `preview: ${unusedNames.slice(0, 5).join(', ')}`,
    ],
    suggested_next_step: '',
  };
}
